File: JINSEOP/프로그래머스_WEEK7/2개 이하로 다른 비트.cpp
#include <string>
#include <vector>
using namespace std;

vector<long long> solution(vector<long long> numbers) {
    vector<long long> answer;

    for (auto num : numbers) {
        if (num % 2 == 0) answer.push_back(num + 1);
        else {
            long long key = 1;

            while (true) {
                if ((num & key) == 0) break;
                key *= 2;
            }
            key /= 2;

            answer.push_back(num + key);
        }
    }

    return answer;
}

// 규칙성

File: DONGUK/프로그래머스_WEEK 7/2개 이하로 다른 비트.py
from collections import deque

def solution(numbers):
    answer = []
    
    for num in numbers:
        bin_num = deque(bin(num)[2:])
        bin_num.appendleft('0')
        
        N = len(bin_num)
        ind = 0
        for i in range(N-1, -1, -1):
            if bin_num[i] == '0':
                ind = i
                break
        
        if bin_num[-1] == '0':
            answer.append(num+1)
        
        
        else:
            bin_num[ind] = '1'
            bin_num[ind+1] = '0'
            
            j = 0
            res = 0
            for i in range(N-1, -1, -1):
                if bin_num[i] == '1':
                    res += 2**j
                j+=1
            answer.append(res)
            
    return answer

